Transcripts filter content

SergioEstevao opened the pull request:
Refs #1848 

This PR  improves the display of the transcript by applying filters to the content, in order to remove Speaker information and better layout the text depending of the existence of paragraphs.

https://github.com/Automattic/pocket-casts-ios/assets/651601/e6b0f8c4-5a56-4431-a117-54bec639630e

## To test

- Start the app
- Ensure that you have the `transcripts` and `newShowNotesEndpoint` FF enabled
- Start playing an episode with Transcript support. Ex: Any episode The Audacity to Podcast
- Open the full screen player
- Tap on the transcript tabs
- See that the layout of text, is respecting paragraphs and no "Speaker" entries or other xml kind of tags `<>` are present on on the text.

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

SergioEstevao committed:
Add filters to clean up and layout transcript content

SergioEstevao committed:
Merge branch 'feature/transcripts_scroll_with_player' into feature/transcripts_filter_content

SergioEstevao committed:
Merge branch 'feature/transcripts_scroll_with_player' into feature/transcripts_filter_content

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

SergioEstevao committed:
Fix spacing.

SergioEstevao commented on podcasts/TranscriptModel.swift:
```diff
@@ -52,7 +52,10 @@ struct TranscriptModel: Sendable {
             let text = cue.text
-            let attributedText = NSAttributedString(string: text + "\n")
+
+            let filteredText: String = ComposeFilter.transcriptFilter.filter(text)
```
In the future ww may want to consider using different filters depending of the subtitle file format.

SergioEstevao committed:
Merge branch 'feature/transcripts_scroll_with_player' into feature/transcripts_filter_content

SergioEstevao committed:
Make fields private

leandroalonso committed:
Rename to TranscriptFilter to avoid confusion with filters

leandroalonso committed:
Add unit test for transcript filter

leandroalonso committed:
Trim string to avoid empty spaces

leandroalonso committed:
Update test name

leandroalonso committed:
Add tests for VTT

leandroalonso committed:
Add tests for tags

leandroalonso commented on podcasts/TranscriptFilter.swift:
```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+protocol TranscriptFilter {
```
I renamed this to `TranscriptFilter` to avoid any confusions with the filters we have in the app.

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/ComposeFilterTests.swift:
```diff
@@ -0,0 +1,71 @@
+
+@testable import podcasts
+
+final class ComposeFilterTests: XCTestCase {
```
I added some tests and I believe we're still quite not there yet. For the transcripts using the `<v person>` the content is not correctly formatted in sentences, it just becomes a huge bulk of text:

<img src="https://github.com/Automattic/pocket-casts-ios/assets/7040243/1ab8d690-9c8a-4477-be58-3f866e67666f" width="300">

SergioEstevao replied:
@leandroalonso I update the filters to better address the layout of text. I also added some unit tests to check VTT and SRT file filtering as a whole to make it easier to spot issues.
Do you mind to give it another look?

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/ComposeFilterTests.swift:
```diff
@@ -0,0 +1,71 @@
+        let expected = """
+        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus imperdiet condimentum ligula quis fringilla. Donec eu ultricies enim. Aenean eu risus leo. Nulla facilisi.
+
+         Mauris ac urna sodales, efficitur ipsum congue, congue arcu. Nam ut lacus eget urna vehicula dictum eget vel mauris.
```
We have some empty spaces added before sentences here (you can see that by deleting an empty space before "Mauris":

```suggestion
        Mauris ac urna sodales, efficitur ipsum congue, congue arcu. Nam ut lacus eget urna vehicula dictum eget vel mauris.
```

leandroalonso reviewed (changes requested):
I've gone ahead and added a few unit tests and did some manual tests. I think we're quite not there yet and we need to refine the solution.

SergioEstevao committed:
Merge branch 'trunk' into feature/transcripts_filter_content

SergioEstevao committed:
Remove spaces after Speaker x:

SergioEstevao committed:
Merge branch 'trunk' into feature/transcripts_filter_content

SergioEstevao committed:
Improve filters to do proper replacements for end of lines with full stop and without full stop

SergioEstevao committed:
Update filter tests to address a single line use.

Filters for transcripts are applied separately for each cue so we shouldn't test for multiple lines/cues

SergioEstevao committed:
Give priority to SRT

SergioEstevao committed:
Just use the same size of font and give preference to New York if available.

SergioEstevao committed:
Improve new lines filters for transcripts

SergioEstevao committed:
Check for question mark too

SergioEstevao committed:
Ensure that question mark and exclamation mark also break line

SergioEstevao committed:
Scroll to range to better location.

SergioEstevao committed:
Merge branch 'trunk' into feature/transcripts_filter_content

SergioEstevao committed:
Fix regex negation

SergioEstevao committed:
Use suggested size

SergioEstevao committed:
Add unit tests for full model formatting

leandroalonso committed:
Merge branch 'trunk' into feature/transcripts_filter_content

leandroalonso commented:
@SergioEstevao FYI: I added a task in #1848 to format HTML transcripts.

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/ComposeFilterTests.swift:
```diff
@@ -0,0 +1,53 @@
+
+        let filtered = ComposeFilter.transcriptFilter.filter(transcript)
+
+        XCTAssertEqual(filtered, "This content should not change ")
```
I see that there's an empty space after the _"change"_ word. I don't think that's a big issue when displaying but the name of the test is misleading.

Maybe we should rename it to "testAddEmptySpaceAfterSentence"?

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/1870/commits/ba7b83d25e4cf2b7c99912567c68d47fa5759dbf

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/ComposeFilterTests.swift:
```diff
@@ -0,0 +1,53 @@
+        let filtered = ComposeFilter.transcriptFilter.filter(transcript)
+
+        let expected = """
+        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
+        Phasellus imperdiet condimentum ligula quis fringilla.
+        Donec eu ultricies enim.
+        Aenean eu risus leo.
+        Nulla facilisi.
```
I feel like breaking sentences based on period might make it harder to read. This is an example, I feel like the original is way easier. 😅

I do understand that a good balance is hard to achieve and on real podcasts it seems ok, so it might be just that this example is really not a good one.

SergioEstevao replied:
I researched to try to find good metrics, we could say for example to do full stop every two sentences or x amount of words but it's heavily influenced by the context of the writing itself.

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/TranscriptModelTests.swift:
```diff
@@ -0,0 +1,124 @@
+
+@testable import podcasts
+
+final class TrasnscriptModelFilterTests: XCTestCase {
```
typo

```suggestion
final class TranscriptModelFilterTests: XCTestCase {
```

SergioEstevao replied:
Fixed on: [Rename file and test](https://github.com/Automattic/pocket-casts-ios/pull/1870/commits/99ca08946ea505c2379e447af7f7335a29d1bdc1)

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/TranscriptModelTests.swift:
```diff
@@ -0,0 +1,124 @@
+
+    func testVTT() throws {
+        let transcript = """
+        WEBVTT - Flying Too High: AI and Air France Flight 447
```
What do you think about changing the words here randomly? To avoid any copyright issues.

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/1870/commits/3bd9fb6c72e069c544fb27de2f991cf11e258632

leandroalonso commented on PocketCastsTests/Tests/Player/Transcripts/TranscriptModelTests.swift:
```diff
@@ -0,0 +1,124 @@
+
+    func testSRT() throws {
+        let transcript = """
+        1
```
Same comment as before (copyright concerns).

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/1870/commits/3bd9fb6c72e069c544fb27de2f991cf11e258632

leandroalonso commented on podcasts/TranscriptFilter.swift:
```diff
@@ -0,0 +1,68 @@
+    static let speakerFilter = RegexFilter(pattern: "Speaker \\d?: *", replacement: "")
+    // Ensure that any full stop starts a new line
+    static let fullStopNewLineFilter = RegexFilter(pattern: "([\\!\\?\\.])\\s*", replacement: "$1\n")
+    // Ensure that end of cues have a space when appended to the next cue
```
If there's no next cue an additional space will be added, right?

SergioEstevao replied:
Correct. I'm assuming the scripts will end in a right way.

leandroalonso commented on podcasts/TranscriptsViewController.swift:
```diff
@@ -230,7 +235,7 @@ class TranscriptsViewController: PlayerItemViewController {
             transcriptView.attributedText = styleText(transcript: transcript, position: position)
             // adjusting the scroll to range so it shows more text
-            let scrollRange = NSRange(location: range.location, length: range.length * 5)
+            let scrollRange = NSRange(location: range.location, length: range.length * 2)
```
Shall we create a task to disable the scrolling/highlight?

SergioEstevao replied:
Will add it.

leandroalonso reviewed (changes requested):
@SergioEstevao tested with multiple podcasts and it's looking good!

Left a few comments, none of them are blockers except the copyright concern I have. It would be good to change the test input for something else.

SergioEstevao committed:
Rename file and test

SergioEstevao committed:
Update sample text to use lorem ipsum translated to english

SergioEstevao committed:
Rename test to better reflect use

SergioEstevao commented:
@leandroalonso thanks for the review, I updated the code according to your suggestions and also changed the sample transcript text on the tests to avoid any copyright issues.

SergioEstevao committed:
Merge branch 'trunk' into feature/transcripts_filter_content
